STM32/I2C/Src/sm4291.c

```diff
@@ -0,0 +1,310 @@
+    return 0;
+}
+
+/*!
+** @brief Get the serial number from the sensor
+*/
+int sm4291GetSerial(sm4291_i2c_handle_t* i2c, uint32_t* result) {
+    uint16_t ser0;
+    uint16_t ser1;
+
+    if(0 != sm4291ReadReg(i2c, ADDR_SER0, &ser0)) {
+        return -1;
+    };
+
+    if(0 != sm4291ReadReg(i2c, ADDR_SER1, &ser1)) {
+        return -2;
+    };
+
+    *result = (((uint32_t)ser1 << 16U) & 0xFFFF0000U) | ((uint32_t)ser0 & 0xFFFFU);
+    return 0;
+}
+
+/*!
+** @brief Get the status register
+*/
+int sm4291GetStatus(sm4291_i2c_handle_t* i2c, uint16_t* status) {
+    return sm4291ReadReg(i2c, ADDR_STATUS, status);
+}
+
+/*!
+** @brief Get the status sync register
+*/
+int sm4291GetStatusSync(sm4291_i2c_handle_t* i2c, uint16_t* status) {
+    return sm4291ReadReg(i2c, ADDR_STATUS_SYNC, status);
+}
+
+/*!
+** @brief Reset the device
+*/
+int sm4291Reset(sm4291_i2c_handle_t* i2c) {
+    return sm4291WriteReg(i2c, ADDR_CMD, CMD_RESET);
+}
+
+/*!
+** @brief Put the device into sleep mode
+**
+** @note The datasheet doesn't specify how t get the device out of sleep mode
+*/
+int sm4291Sleep(sm4291_i2c_handle_t* i2c) {
+    return sm4291WriteReg(i2c, ADDR_CMD, CMD_SLEEP);
+}
+
+/***************************************************************************************************
+** PRIVATE FUNCTIONS
+***************************************************************************************************/
+
+/*!
+** @brief Reads a 16-bit register from the device
+*/
+static int sm4291ReadReg(sm4291_i2c_handle_t* i2c, uint8_t addr, uint16_t* result) {
+    assert_param(i2c);
+
+    uint16_t dev_addr = i2c->crc ? TEMP_I2C_CRC_ADDR : TEMP_I2C_NO_CRC_ADDR;
+
+    if(i2c->crc) {
+        uint8_t crc4_data[2U] = {(uint8_t)(addr >> 4U), ((addr & 0xFU) << 4U) & 0x1U};
+        uint8_t crc4 = genCrc4(crc4_data, 2U);
+
+        uint8_t buf[2U] = {addr, (uint8_t)(0x10U | crc4)};
+        if(HAL_OK != HAL_I2C_Master_Transmit(i2c->i2c, dev_addr << 1, buf, 2, 1)) {
+            return -1;
+        };
+
+        uint8_t buf1[3U] = {0};
+        if(HAL_OK != HAL_I2C_Master_Receive(i2c->i2c, dev_addr << 1, buf1, 3, 1)) {
+            return -2;
+        };
+
+        if(genCrc8(buf1, 2) != buf1[2]) {
+
+            return -3;
+        }
+
+        *result = ((uint16_t) buf1[1] << 8U) | buf1[0];
+    }
+    else {
+        if(HAL_OK != HAL_I2C_Master_Transmit(i2c->i2c, dev_addr << 1, &addr, 1, 1)) {
+            return -1;
+        };
+
+        if(HAL_OK != HAL_I2C_Master_Receive(i2c->i2c, dev_addr << 1, (uint8_t*)result, 2, 1)) {
+            return -2;
+        };
+    }
+
+    return 0;
+}
+
+/*!
+** @brief Writes a 16-bit register to the device
+*/
+static int sm4291WriteReg(sm4291_i2c_handle_t* i2c, uint8_t addr, uint16_t reg) {
+    assert_param(i2c);
+
+    uint16_t dev_addr = i2c->crc ? TEMP_I2C_CRC_ADDR : TEMP_I2C_NO_CRC_ADDR;
+    if(i2c->crc) {
+        uint8_t crc4_data[2U] = {(uint8_t)(addr >> 4U), ((addr & 0xF) << 4U) & 0x1U};
```

Comment:
Also here I would prefer addr -> reg and reg -> value/update_val, but just preference